### PR TITLE
1 - Move iio files to libraries

### PR DESCRIPTION
**Pull request by mchindri**
- Copy iio source files to libraries/iio
- Add libraries/iio/libtinyiiod as submodule

This change will be done in order to build iio as a library because tinyiiod is used only by iio so for me it makes sense both to be in one place.
More, with this approach new project will not depend on defined symbols needed by tinyiiod.
Finally, doing this will help speed up the developement of new changes that have been decide to be done on iio and there is no need to update all projects on each change which would slow the development. When all new changes are stable, the project can be modified to use the new iio library with small changes.
During development only iio_demo project will be modified in order for the changes to be better understood.

### CODE REVIEW

File: libraries/iio/Makefile [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+OBJS = $(SRCS:.c=.o)
+
+CFLAGS += $(IIO_DEFS) \
+	  -I"$(TINYIIOD_DIR)"
+
+.PHONEY = all clean
+
+#Init tiniiod submodule if not initialized
+ifeq ($(wildcard $(TINYIIOD_DIR)/README.md),)
```
Comment (scuciurean) [inline libraries/iio/Makefile:40]: Wildcards are used for a specific pattern, not for a single file

File: libraries/iio/Makefile [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+all: $(LIB_NAME)
+endif
+
+$(LIB_NAME): $(OBJS)
+	$(AR) $(ARFLAGS) $@ $(OBJS)
+
+clean:
+	-$(call remove_fun,$(OBJS) $(LIB_NAME))
+
+AUX = $(wildcard $(TINYIIOD_DIR)/README.md)
```
Comment (scuciurean) [inline libraries/iio/Makefile:54]: Same here

File: libraries/iio/iio.c [hunk 3]
```diff
@@ -0,0 +1,843 @@
+/***************************************************************************//**
```
Comment (amiclaus) [inline libraries/iio/iio.c:1]: do we want to have duplicate files?
https://github.com/analogdevicesinc/no-OS/blob/master/iio/iio.c

File: libraries/iio/iio.h [hunk 4]
```diff
@@ -0,0 +1,98 @@
+/***************************************************************************//**
```
Comment (amiclaus) [inline libraries/iio/iio.h:1]: do we want to have duplicate files?
https://github.com/analogdevicesinc/no-OS/blob/master/iio/iio.h

File: libraries/iio/iio_types.h [hunk 5]
```diff
@@ -0,0 +1,119 @@
+/***************************************************************************//**
```
Comment (amiclaus) [inline libraries/iio/iio_types.h:1]: same as above

**Comment by amiclaus** [comment]
With this PR you add: `libraries/iio/libtinyiiod` as submodule. 
The submodule already exists under `libraries` therefore we will have 2x libtinyiiod submodules in the same repo (no-OS).

File: libraries/iio/Makefile (hunk 1 above)
Comment (mchindri) [reply libraries/iio/Makefile:40]: In this case wildcard is used to check if file exists.

File: libraries/iio/Makefile (hunk 2 above)
Comment (mchindri) [reply libraries/iio/Makefile:54]: This was for testing, I will remove it

File: libraries/iio/Makefile (hunk 1 above)
Comment (cuciureansergiu) [reply libraries/iio/Makefile:40]: > In this case wildcard is used to check if file exists.

You can use this as an example

FILE=/etc/resolv.conf
if [ -f "$FILE" ]; then
    echo "$FILE exists."
else 
    echo "$FILE does not exist."
fi

Also wouldn't assume if a file named README.md exists in a folder, it world be a repository

Comment (mchindri) [reply libraries/iio/Makefile:40]: I think is simpler this way.
I didn't find an other way to check if the submodule is initialized.

Comment (mchindri) [reply libraries/iio/Makefile:40]: Also it seems what you propose is a shell scrips, which I would rather avoid inside the makefile

File: libraries/iio/iio.c (hunk 3 above)
Comment (mchindri) [reply libraries/iio/iio.c:1]: yes, the motivation is in the description

**Comment by mchindri** [comment]
v2:
 - Remove last commit from this PR. Will create an other PR for it

File: libraries/iio/Makefile (hunk 1 above)
Comment (adrimbarean) [reply libraries/iio/Makefile:40]: You could check for the .git folder. This might be more generic.

File: libraries/iio/Makefile [hunk 6]
```diff
@@ -0,0 +1,56 @@
...
+ifeq ($(wildcard $(TINYIIOD_DIR)/README.md),)
+all:
+	git submodule update --init --remote -- $(TINYIIOD_DIR)
+	$(MAKE) $(LIB_NAME)
+else
+all: $(LIB_NAME)
+endif
+
+$(LIB_NAME): $(OBJS)
+	$(AR) $(ARFLAGS) $@ $(OBJS)
```
Comment (adrimbarean) [inline libraries/iio/Makefile:49]: nitpick: this could also be possible and a bit more concise
```suggestion
	$(AR) $(ARFLAGS) $@ $^
```

**Comment by mchindri** [comment]
v3:
 - Move iio files to libraries/iio with git mv and then copy them back to iio. This way changes can be tracked in libraries/iio
 - Remove Makefile commit  from this pull request. Add it here https://github.com/analogdevicesinc/no-OS/pull/674

File: libraries/iio/Makefile (hunk 6 above)
Comment (mchindri) [reply libraries/iio/Makefile]: will solve here: https://github.com/analogdevicesinc/no-OS/pull/674

File: libraries/iio/Makefile (hunk 2 above)
Comment (mchindri) [reply libraries/iio/Makefile]: will solve here: https://github.com/analogdevicesinc/no-OS/pull/674

File: libraries/iio/Makefile (hunk 1 above)
Comment (mchindri) [reply libraries/iio/Makefile]: will solve here: https://github.com/analogdevicesinc/no-OS/pull/674

**Pull request merged**